File: IT/1/Containers/Java/SetTheory.java
/* file SetTheory.java */

import java.util.SortedSet;
import java.util.TreeSet;

public class SetTheory {
	public static void main(String [] args) {
		/* dichiara due Set di numeri interi */

		SortedSet<Integer> a = new TreeSet<Integer>();

		a.add(0);
		a.add(1);
		a.add(2);

		SortedSet<Integer> b = new TreeSet<Integer>();

		b.add(2);
		b.add(3);
		b.add(4);

		System.out.println("A = " + a);
		System.out.println("B = " + b);

		/* unione di A e B */

		SortedSet<Integer> s_union = new TreeSet<Integer>(a);
		s_union.addAll(b);

		System.out.println("A u B = " + s_union);

		/* intersezione di A e B */

		SortedSet<Integer> intersection = new TreeSet<Integer>(a);
		intersection.retainAll(b);

		System.out.println("A & B = " + intersection);

		/* differenza tra A e B */

		SortedSet<Integer> diff = new TreeSet<Integer>(a);
		diff.removeAll(b);
		System.out.println("A - B = " + diff);

		/* ...e differenza tra B e A */

		diff = new TreeSet<Integer>(b);
		diff.removeAll(a);
		System.out.println("B - A = " + diff);

		/* differenza simmetrica tra A e B */

		SortedSet<Integer> s_diff = new TreeSet<Integer>(a);

		/* calcola l'unione A u B */

		s_diff.addAll(b);

		/* calcola la differenza simmetrica A ^ B  = (A u B) - (A & B) */

		s_diff.removeAll(intersection);

		System.out.println("A ^ B = " + s_diff);
	}
}

File: IT/1/Containers/C++/setTheory.cpp
/* 
file setTheory.cpp
compilare con: g++ -std=c++11 setTheory.cpp
*/

#include <iostream>
#include <iterator>
#include <algorithm>
#include <set>

using std::cout;
using std::endl;
using std::ostream_iterator;
using std::set;

int main()
{
	/* dichiara due insiemi di numeri interi */

	set<int> a = {0, 1, 2};
	set<int> b = {2, 3, 4};

	cout << "A = [";
	copy(a.begin(), a.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	cout << "B = [";
	copy(b.begin(), b.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* unione di A e B */

	set<int> s_union;
	set_union(a.begin(), a.end(), b.begin(), b.end(), inserter(s_union, s_union.begin()));

	cout << "A u B = [";
	copy(s_union.begin(), s_union.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* intersezione di A e B */

	set<int> intersection;
	set_intersection(a.begin(), a.end(), b.begin(), b.end(), inserter(intersection, intersection.begin()));

	cout << "A & B = [";
	copy(intersection.begin(), intersection.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* differenza tra A e B */

	set<int> diff;
	set_difference(a.begin(), a.end(), b.begin(), b.end(), inserter(diff, diff.begin()));

	cout << "A - B = [";
	copy(diff.begin(), diff.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* ...e differenza tra B e A */

	diff.clear();
	set_difference(b.begin(), b.end(), a.begin(), a.end(), inserter(diff, diff.begin()));

	cout << "B - A = [";
	copy(diff.begin(), diff.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;

	/* differenza simmetrica tra A e B */

	set<int> s_diff;
	set_symmetric_difference(a.begin(), a.end(), b.begin(), b.end(), inserter(s_diff, s_diff.begin()));

	cout << "A ^ B = [";
	copy(s_diff.begin(), s_diff.end(), ostream_iterator<int>(cout, ", "));
	cout << "\b\b]" << endl;
}

